books/RayTracingTheNextWeek.html

```diff
@@ -2057,105 +2057,147 @@
+            std::cerr << "STBI: " << stbi_failure_reason() << '\n';
         }
 
-        virtual color value(double u, double v, const vec3& p) const override {
-            // If we have no texture data, then return solid cyan as a debugging aid.
-            if (data == nullptr)
-                return color(0,1,1);
+        ~rtw_image() { STBI_FREE(data); }
 
-            // Clamp input texture coordinates to [0,1] x [1,0]
-            u = interval(0,1).clamp(u);
-            v = 1.0 - interval(0,1).clamp(v);  // Flip V to image coordinates
+        bool load(const std::string filename) {
+            // Loads image data from the given file name. Returns true if the load succeeded.
+            auto components_per_pixel = bytes_per_pixel; // Components per pixel
+            data = stbi_load(filename.c_str(),
+                &image_width, &image_height, &components_per_pixel, bytes_per_pixel);
+            bytes_per_scanline = image_width * bytes_per_pixel;
+            return data != nullptr;
+        }
 
-            auto i = static_cast<int>(u * width);
-            auto j = static_cast<int>(v * height);
+        int width()  const { return (data == nullptr) ? 0 : image_width; }
+        int height() const { return (data == nullptr) ? 0 : image_height; }
 
-            // Clamp integer mapping, since actual coordinates should be less than 1.0
-            if (i >= width)  i = width-1;
-            if (j >= height) j = height-1;
+        const unsigned char* pixel_data(int x, int y) const {
+            // Return the address of the three bytes of the pixel at x,y (or magenta if no data).
+            static unsigned char magenta[] = { 255, 0, 255 };
+            if (!data) return magenta;
 
-            const auto color_scale = 1.0 / 255.0;
-            auto pixel = data + j*bytes_per_scanline + i*bytes_per_pixel;
+            // Clamp pixel coordintes to [0,width-1] x [0,height-1]
+            x = (x < 0) ? 0 : (x < image_width)  ? x : (image_width  - 1);
+            y = (y < 0) ? 0 : (y < image_height) ? y : (image_height - 1);
 
-            return color(color_scale*pixel[0], color_scale*pixel[1], color_scale*pixel[2]);
+            return data + y*bytes_per_scanline + x*bytes_per_pixel;
         }
 
       private:
+        const int bytes_per_pixel = 3;
         unsigned char *data;
-        int width, height;
+        int image_width, image_height;
         int bytes_per_scanline;
     };
 
+    // Restore MSVC compiler warnings
+    #ifdef _MSC_VER
+        #pragma warning (pop)
+    #endif
+
+    #endif
     ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
-    [Listing [img-texture]: <kbd>[texture.h]</kbd> Image texture class]
+    [Listing [rtw_image]: <kbd>[rtw_stb_image.h] The rtw_image helper class]
+
+The preceding listing assumes that you have copied the `stb_image.h` header into a folder called
+`external`. Adjust according to your directory structure.
```

Comment:
Consider moving this descriptor to above the source block.

Reply:
Sounds good.